File: hash-map-set-table-(hashing)/347.py
# Tc = O(n * logk)
# Sc = O(n)

from collections import defaultdict
from typing import List
class Solution:
    def topKFrequentElement(self, nums: List[int], k : int) -> List[int]:
        d = defaultdict(int)
        for num in nums:
            d[num] += 1
        heap = []
        for key, val in d.items():
            if len(heap) < k or val > heap[0][0]:
                heapq.heappush(heap, [val, key])
            if len(heap) > k:
                heapq.heappop(heap)
        return [i[1] for i in heap]
        

File: hash-map-set-table-(hashing)/347.js
//  Tc = O(n * logk)
//  Sc = O(n)
var topKFrequentElement = function(nums, k){
    const freqMap = new Map()
    for(const num of nums){
        freqMap.set(num, (freqMap.get(num) || 0) +1)
    }
    const heap = []
    for(const [num, freq] of freqMap.entries()){
        if(heap.length < k){
            heap.push([freq, num])
            heap.sort((a, b) => a[0] - b[0])
        }else if(freq > heap[0][0]){
            heap[0] = [freq, num]
            heap.sort((a, b) => a[0] - b[0])
        }
    }
    return heap.map(item => item[1])
}